### PR TITLE
Support to fetch vSphere and k8s Version from k8s cluster and fetch corresponding Deployment Yamls

### PR DESCRIPTION
**Pull request by ridaz**
**This PR brings in the following support :**
1. Fetch vSphere and k8s version dynamically
2. Fetch CPI and CSI deployment yamls for corresponding vSphere and k8s version from the compatibility matrix
3. Unit tests for different cases of versions/yamls.

**Ran the unit tests :** 

```
?       github.com/vmware-tanzu/vsphere-kubernetes-drivers-operator     [no test files]
?       github.com/vmware-tanzu/vsphere-kubernetes-drivers-operator/api/v1alpha1        [no test files]
ok      github.com/vmware-tanzu/vsphere-kubernetes-drivers-operator/controllers 24.108s coverage: 32.4% of statements
ok      github.com/vmware-tanzu/vsphere-kubernetes-drivers-operator/pkg/client  21.424s coverage: 60.0% of statements
?       github.com/vmware-tanzu/vsphere-kubernetes-drivers-operator/pkg/context [no test files]
ok      github.com/vmware-tanzu/vsphere-kubernetes-drivers-operator/pkg/drivers/cpi     2.199s  coverage: 88.0% of statements
ok      github.com/vmware-tanzu/vsphere-kubernetes-drivers-operator/pkg/drivers/csi     1.066s  coverage: 77.1% of statements
?       github.com/vmware-tanzu/vsphere-kubernetes-drivers-operator/pkg/models  [no test files]
ok      github.com/vmware-tanzu/vsphere-kubernetes-drivers-operator/pkg/session 1.843s  coverage: 61.2% of statements

```

**Compatibility Matrix can be referred as :**

```
{
    "CSI" : {
            "2.2.1" : {
                    "vSphere" : { "min" : "6.7", "max": "7.0"},
                    "k8s" : {"min": "1.18", "max": "1.21"},
                    "isCPIRequired" : false,
                    "deploymentPath": [
                    "https://raw.githubusercontent.com/kubernetes-sigs/vsphere-csi-driver/v2.2.1/manifests/v2.2.1/rbac/vsphere-csi-controller-rbac.yaml",
                    "https://raw.githubusercontent.com/kubernetes-sigs/vsphere-csi-driver/v2.2.1/manifests/v2.2.1/rbac/vsphere-csi-node-rbac.yaml",
                    "https://raw.githubusercontent.com/kubernetes-sigs/vsphere-csi-driver/v2.2.1/manifests/v2.2.1/deploy/vsphere-csi-controller-deployment.yaml",
                    "https://raw.githubusercontent.com/kubernetes-sigs/vsphere-csi-driver/v2.2.1/manifests/v2.2.1/deploy/vsphere-csi-node-ds.yaml"]
                    }
        },
    "CPI" : {
            "1.20.0" : {
                    "vSphere" : { "min" : "6.7", "max": "7.0"},
                    "k8s" : {"skewVersion": "1.21"},
                    "deploymentPath": [
                    "https://raw.githubusercontent.com/kubernetes/cloud-provider-vsphere/v1.20.0/manifests/controller-manager/cloud-controller-manager-roles.yaml",
                    "https://raw.githubusercontent.com/kubernetes/cloud-provider-vsphere/v1.20.0/manifests/controller-manager/cloud-controller-manager-role-bindings.yaml",
                    "https://raw.githubusercontent.com/kubernetes/cloud-provider-vsphere/v1.20.0/manifests/controller-manager/vsphere-cloud-controller-manager-ds.yaml"]
                    }
        }
             
}

```

```
I0902 09:53:44.830374       1 vdoconfig_controller.go:1096] controllers/VDOConfig "msg"="vSphere Versions "  "version"=["7.0.3"]
I0902 09:53:44.830573       1 vdoconfig_controller.go:1097] controllers/VDOConfig "msg"="k8s Versions "  "version"="1.22"
I0902 09:53:44.830784       1 vdoconfig_controller.go:1125] controllers/VDOConfig "msg"="Corresponding CSI Version "  "version"="2.2.1"
I0902 09:53:44.830900       1 vdoconfig_controller.go:1145] controllers/VDOConfig "msg"="vSphere Versions "  "version"=["7.0.3"]
I0902 09:53:44.830917       1 vdoconfig_controller.go:1146] controllers/VDOConfig "msg"="k8s Versions "  "version"="1.22"
I0902 09:53:44.830935       1 vdoconfig_controller.go:1171] controllers/VDOConfig "msg"="Corresponding CPI Version "  "version"="1.20.0"
I0902 09:53:44.831055       1 vdoconfig_controller.go:159] controllers/VDOConfig "msg"="CSI deployment yamls : "  "list"=["https://raw.githubusercontent.com/kubernetes-sigs/vsphere-csi-driver/v2.2.1/manifests/v2.2.1/rbac/vsphere-csi-controller-rbac.yaml","https://raw.githubusercontent.com/kubernetes-sigs/vsphere-csi-driver/v2.2.1/manifests/v2.2.1/rbac/vsphere-csi-node-rbac.yaml","https://raw.githubusercontent.com/kubernetes-sigs/vsphere-csi-driver/v2.2.1/manifests/v2.2.1/deploy/vsphere-csi-controller-deployment.yaml","https://raw.githubusercontent.com/kubernetes-sigs/vsphere-csi-driver/v2.2.1/manifests/v2.2.1/deploy/vsphere-csi-node-ds.yaml"]
I0902 09:53:44.831399       1 vdoconfig_controller.go:160] controllers/VDOConfig "msg"="CPI deployment yamls : "  "list"=["https://raw.githubusercontent.com/kubernetes/cloud-provider-vsphere/v1.20.0/manifests/controller-manager/cloud-controller-manager-roles.yaml","https://raw.githubusercontent.com/kubernetes/cloud-provider-vsphere/v1.20.0/manifests/controller-manager/cloud-controller-manager-role-bindings.yaml","https://raw.githubusercontent.com/kubernetes/cloud-provider-vsphere/v1.20.0/manifests/controller-manager/vsphere-cloud-controller-manager-ds.yaml"]
```

```
kubectl get pods -A

NAMESPACE           NAME                                     READY   STATUS    RESTARTS   AGE
kube-system         vsphere-cloud-controller-manager-dfl97   1/1     Running   0          3m16s
kube-system         vsphere-csi-controller-55f949b89-fvvjv   6/6     Running   0          2m59s
kube-system         vsphere-csi-node-f79nl                   3/3     Running   0          2m58s
kube-system         vsphere-csi-node-vn57x                   3/3     Running   0          2m59s
vmware-system-vdo   vdo-controller-manager-d57b9684d-ggt6t   2/2     Running   0          3m42s

```
```

kubectl describe nodes | grep ProviderID
ProviderID:                   vsphere://423fa6d2-76c2-c6c0-cd3a-5e98ead5d42a
ProviderID:                   vsphere://423fd8bc-d5d4-c071-3055-52cae661ef83
```

```
kubectl get vdoconfig -A -o yaml
apiVersion: v1
items:
- apiVersion: vdo.vmware.com/v1alpha1
  kind: VDOConfig
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"vdo.vmware.com/v1alpha1","kind":"VDOConfig","metadata":{"annotations":{},"name":"vdoconfig-sample","namespace":"vmware-system-vdo"},"spec":{"cloudProvider":{"vsphereCloudConfigs":["vspherecloudconfig-sample"]},"storageProvider":{"vsphereCloudConfig":"vspherecloudconfig-sample"}}}
    creationTimestamp: "2021-09-01T21:19:55Z"
    generation: 1
    name: vdoconfig-sample
    namespace: vmware-system-vdo
    resourceVersion: "87264"
    uid: 4691f655-e76d-4518-8309-968f86c3d8eb
  spec:
    cloudProvider:
      vsphereCloudConfigs:
      - vspherecloudconfig-sample
    storageProvider:
      vsphereCloudConfig: vspherecloudconfig-sample
  status:
    cpi:
      'nodeStatus ':
        master01: ready
        worker01: ready
      phase: Configured
    csi:
      phase: Configuring
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""

```

```
kubectl describe pvc example-vanilla-block-pvc
Name:          example-vanilla-block-pvc
Namespace:     default
StorageClass:  k8s-default-storage
Status:        Bound
Volume:        pvc-f8f9fce8-ef58-4393-bfe0-56a03e662178
Labels:        <none>
Annotations:   pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: csi.vsphere.vmware.com
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      2Gi
Access Modes:  RWO
VolumeMode:    Filesystem
Used By:       <none>
Events:
  Type    Reason                 Age   From                                                                                                Message
  ----    ------                 ----  ----                                                                                                -------
  Normal  ExternalProvisioning   66s   persistentvolume-controller                                                                         waiting for a volume to be created, either by external provisioner "csi.vsphere.vmware.com" or manually created by system administrator
  Normal  Provisioning           66s   csi.vsphere.vmware.com_vsphere-csi-controller-55f949b89-fvvjv_cf28ca96-ca83-4f96-ba5d-e6462c0e6a42  External provisioner is provisioning volume for claim "default/example-vanilla-block-pvc"
  Normal  ProvisioningSucceeded  62s   csi.vsphere.vmware.com_vsphere-csi-controller-55f949b89-fvvjv_cf28ca96-ca83-4f96-ba5d-e6462c0e6a42  Successfully provisioned volume pvc-f8f9fce8-ef58-4393-bfe0-56a03e662178


```

### CODE REVIEW

**Review by jvrahav — CHANGES_REQUESTED** [review]
can we move this to a function say fetch CheckCompatAndRetrieveSpec

File: controllers/vdoconfig_controller.go [hunk 1]
```diff
@@ -116,6 +119,19 @@ func (r *VDOConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 		return ctrl.Result{}, err
 	}
 
+	discoveryClient, _ := discovery.NewDiscoveryClientForConfig(r.ClientConfig)
+	k8sServerVersion, err := discoveryClient.ServerVersion()
```
Comment (jvrahav) [inline controllers/vdoconfig_controller.go:123]: Can we move this to a function say FetchCompatibleVersions

File: controllers/vdoconfig_controller.go [hunk 2]
```diff
@@ -129,8 +145,16 @@ func (r *VDOConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
...
+			vdoctx.Logger.Error(err, "Error occurred when fetching the CSI deployment yamls")
+			return ctrl.Result{}, err
+		}
+		cpiDeploymentYamls, err = r.fetchCpiDeploymentYamls(vdoctx, matrix, vSphereVersions, k8sServerVersion.Major, k8sServerVersion.Minor)
+		if err != nil {
+			vdoctx.Logger.Error(err, "Error occurred when fetching the CPI deployment yamls")
+			return ctrl.Result{}, err
+		}
 
 		vdoctx.Logger.V(4).Info("CSI deployment yamls : ", "list", csiDeploymentYamls)
```
Comment (jvrahav) [inline controllers/vdoconfig_controller.go:159]: we can avoid logging the url of spec files

File: controllers/vdoconfig_controller.go [hunk 3]
```diff
@@ -197,6 +221,81 @@ func (r *VDOConfigReconciler) fetchVsphereCloudConfigItems(vdoctx vdocontext.VDO
 	return vsphereCloudConfigItems, nil
 }
 
+func (r *VDOConfigReconciler) fetchVsphereVersions(vdoctx vdocontext.VDOContext, req ctrl.Request, vdoConfig *vdov1alpha1.VDOConfig) ([]string, error) {
+	vsphereCloudConfigsList := vdoConfig.Spec.CloudProvider.VsphereCloudConfigs
+	// if len of cloudProvider is > 0 , don't get storage provider
+	if len(vsphereCloudConfigsList) <= 0 {
+		vdoctx.Logger.V(4).Info("CPI is not configured for VDO")
+		vsphereCloudConfigsList = append(vsphereCloudConfigsList, vdoConfig.Spec.StorageProvider.VsphereCloudConfig)
```
Comment (jvrahav) [inline controllers/vdoconfig_controller.go:229]: If CPI is not configured would this cause a NIL reference exception ???

File: controllers/vdoconfig_controller.go [hunk 4]
```diff
@@ -901,44 +1000,177 @@ func (r *VDOConfigReconciler) reconcileCSISecret(ctx vdocontext.VDOContext, conf
 	return config, nil
 }
 
-func (r *VDOConfigReconciler) compareVersions(currentVersion, latestVersion string) (version string) {
+// compareVersions checks if the given version lies between min and max version
+func (r *VDOConfigReconciler) compareVersions(minVersion, currentVersion, maxVersion string) (bool, error) {
+	minVersionList := strings.Split(minVersion, ".")
```
Comment (jvrahav) [inline controllers/vdoconfig_controller.go:1005]: seems https://github.com/hashicorp/go-version would be a better fit here to compare versions

File: controllers/vdoconfig_controller.go [hunk 5]
```diff
@@ -901,44 +1000,177 @@ func (r *VDOConfigReconciler) reconcileCSISecret(ctx vdocontext.VDOContext, conf
...
-	// Currently fetching the latest version of CSI and corresponding Deployment Yamls
-	// TODO Fetch vsphere and k8s version from the cluster and select suitable CSI version
+// compareSkewVersions checks if the given version matches with the skew version
+func (r *VDOConfigReconciler) compareSkewVersions(currentVersion, supportedVersion string) (bool, error) {
+
+	currentVersionList := strings.Split(currentVersion, ".")
+	supportedVersionList := strings.Split(supportedVersion, ".")
+
+	for x := range supportedVersionList {
+		if (strings.Contains(supportedVersionList[x], "x")) || (strings.Contains(supportedVersionList[x], "X")) {
```
Comment (jvrahav) [inline controllers/vdoconfig_controller.go:1068]: please explore using https://github.com/hashicorp/go-version for version comparisons

**Review by jvrahav — CHANGES_REQUESTED** [review]

File: controllers/vdoconfig_controller.go [hunk 6]
```diff
@@ -197,7 +189,96 @@ func (r *VDOConfigReconciler) fetchVsphereCloudConfigItems(vdoctx vdocontext.VDO
...
+			return nil, errors.Wrapf(err, "unable to verify session for vc %s", vcIp)
+		}
+
+		r.Logger.V(4).Info("verified vc session", "isActive", state)
+
+		config.Status.Config = vdov1alpha1.VsphereConfigVerified
+		config.Status.Message = ""
+	}
+	return sess, nil
+}
```
Comment (aggarwalta) [inline controllers/vdoconfig_controller.go:279]: This code seems redundant with the already existing function reconcileVCCredentials. Can we try to refactor this?

File: controllers/vdoconfig_controller.go [hunk 7]
```diff
@@ -197,7 +189,96 @@ func (r *VDOConfigReconciler) fetchVsphereCloudConfigItems(vdoctx vdocontext.VDO
...
+	}
+
+	vcIp := config.Spec.VcIP
+	sess, err := session.GetOrCreate(ctx, vcIp, datacenter, vcUser, vcUserPwd, config.Spec.Thumbprint)
+	if err != nil {
+		config.Status.Config = vdov1alpha1.VsphereConfigFailed
+		config.Status.Message = fmt.Sprintf("Error establishing session with vcenter %s for user %s", vcIp, vcUser)
+		return nil, errors.Wrapf(err, "Error establishing session with vcenter %s for user %s", vcIp, vcUser)
+	}
+
```
Comment (jvrahav) [inline controllers/vdoconfig_controller.go:264]: we should just return session here. the rest of the code in the function isnt required

File: controllers/vdoconfig_controller.go [hunk 8]
```diff
@@ -197,7 +189,96 @@ func (r *VDOConfigReconciler) fetchVsphereCloudConfigItems(vdoctx vdocontext.VDO
...
+	vsphereCloudConfigsList = vdoConfig.Spec.CloudProvider.VsphereCloudConfigs
+
+	if len(vsphereCloudConfigsList) <= 0 {
+		vsphereCloudConfigsList = []string{vdoConfig.Spec.StorageProvider.VsphereCloudConfig}
+	}
+	var vsphereCloudConfigItems []vdov1alpha1.VsphereCloudConfig
+	for _, vsphereCloudConfig := range vsphereCloudConfigsList {
+		vsphereCloudConfigItem, err := r.fetchVSphereCloudConfig(vdoctx, vsphereCloudConfig, req.Namespace)
+		vsphereCloudConfigItems = append(vsphereCloudConfigItems, *vsphereCloudConfigItem)
+		if err != nil {
```
Comment (jvrahav) [inline controllers/vdoconfig_controller.go:214]: check for error before appending to list

File: controllers/vdoconfig_controller.go [hunk 9]
```diff
@@ -901,44 +982,170 @@ func (r *VDOConfigReconciler) reconcileCSISecret(ctx vdocontext.VDOContext, conf
...
+	}
+
+	ctx.Logger.V(4).Info("Corresponding CPI Version ", "version", cpiVersion)
+
+	r.CpiDeploymentYamls = matrix.CPISpecList[cpiVersion].DeploymentPaths
+
+	return nil
+}
+
+func (r *VDOConfigReconciler) CheckCompatAndRetrieveSpec(matrixConfigUrl string, ctx vdocontext.VDOContext, req ctrl.Request, vdoConfig *vdov1alpha1.VDOConfig) error {
```
Comment (jvrahav) [inline controllers/vdoconfig_controller.go:1120]: use context as first parameter

File: controllers/vdoconfig_controller.go [hunk 10]
```diff
@@ -901,44 +982,170 @@ func (r *VDOConfigReconciler) reconcileCSISecret(ctx vdocontext.VDOContext, conf
...
+			}
+
+			if isVsphereVersion && isK8sVersion {
+				cpiVersion = versionList[v]
+				break
+			}
+		}
+	}
+
+	if cpiVersion == "" || len(cpiVersion) <= 0 {
```
Comment (jvrahav) [inline controllers/vdoconfig_controller.go:1109]: check for length is sufficient

File: controllers/vdoconfig_controller.go [hunk 11]
```diff
@@ -901,44 +982,170 @@ func (r *VDOConfigReconciler) reconcileCSISecret(ctx vdocontext.VDOContext, conf
...
+			}
+
+			if isVsphereVersion && isK8sVersion {
+				csiVersion = versionList[v]
+				break
+			}
+		}
+	}
+
+	if csiVersion == "" || len(csiVersion) <= 0 {
```
Comment (jvrahav) [inline controllers/vdoconfig_controller.go:1065]: check for length is sufficient

**Review by jvrahav — CHANGES_REQUESTED** [review]

File: controllers/vdoconfig_controller.go [hunk 12]
```diff
@@ -197,7 +193,83 @@ func (r *VDOConfigReconciler) fetchVsphereCloudConfigItems(vdoctx vdocontext.VDO
...
+		err := r.Get(ctx, key, vcCredsSecret)
+		if err != nil {
+			config.Status.Config = vdov1alpha1.VsphereConfigFailed
+			config.Status.Message = fmt.Sprintf("could not fetch vc credentials secret %s", config.Spec.Credentials)
+			return nil, errors.Wrapf(err, "could not fetch vc credentials secret %s", config.Spec.Credentials)
+		}
+
+		vcUser = string(vcCredsSecret.Data["username"])
+		vcUserPwd = string(vcCredsSecret.Data["password"])
+	}
```
Comment (aggarwalta) [inline controllers/vdoconfig_controller.go:255]: We can use the existing function **fetchVcCredentials** for this.

**Review by jvrahav — APPROVED** [review]
Thanks for making the changes and adding the tests

**Review by aggarwalta — APPROVED** [review]
LGTM, thanks!

**Review by asifdxtreme — APPROVED** [review]
LGTM

**Pull request merged**